### DISCUSSION TITLE
til-site/posts/the-unreasonable-effectiveness-of-vms-in-hacker-pedagogy/

### DISCUSSION THREAD

**Discussion by giscus** [hiAndrewQuinn/til#11] [Announcements]
# til-site/posts/the-unreasonable-effectiveness-of-vms-in-hacker-pedagogy/

Here’s a secret. If you have Vagrant and VirtualBox installed, and your colleague does too, then you can both bring up an near-totally identical blank slate Debian 12 Linux VM by running
1 2 3 4 5 6 mkdir tutorial/ cd tutorial/ vagrant init debian/bookworm64 vagrant up vagrant ssh . This works regardless of whether you or they are on Linux, Mac1, BSD, or even Windows. (Through the magic of aliasing, mkdir and cd even work in PowerShell.

https://hiandrewquinn.github.io/til-site/posts/the-unreasonable-effectiveness-of-vms-in-hacker-pedagogy/

**Comment by pmarreck** [comment]
I swear that there should be a tag called #PeopleWhoDontKnowAboutNix

Every single developer on my team (we're 4 in total now) has exactly the same environment thanks to a single flake.nix file that I wrote. (And it wasn't that hard. Certainly easier than wasting time troubleshooting individual dev's configuration differences. And certainly less frustrating than dealing with Docker.) We haven't had a single misconfiguration problem in months now. And if any of us updates the flake.nix file (which then updates the flake.lock file), it automatically updates everyone's environment the next time they pull that file down thanks to direnv's integration with Nix. It's magical, and it entirely solves this problem. And because it's all declarative, there's no "running the same sequence of commands to get to the same configuration" imperatively. Look into it.

**Reply by hiAndrewQuinn** [reply]
Ah, but I'm not trying to write a single, unified flake for my team. I'm trying to write tutorials for random people on the Internet, people who may be very very distant from me in space, time, technical ability level, ability/willingness to communicate with me directly, and actual tech stack.

Part of what I want to do in a tutorial is explicitly communicate "here are the dependencies *this* thing you're trying to do needs." To someone learning how to run the system, running the commands themselves is a feature, not a bug.

I say all this as a fan of the Nix and Guix projects. But if you're good enough to be running Nix, you're probably good enough to skip the sausage-making part of the tutorial. That is a low percentage of the tutorial-seeking population.

**Reply by pmarreck** [reply]
Makes sense! And the dependency story (whatever solution you end up using) seems to have certainly gotten more complicated over the years, perhaps because more projects simply depend on "more of the shoulders of past giants to stand on"

**Reply by hiAndrewQuinn** [reply]
This reminds me, now that I have a functioning mini PC running Proxmox in my homelab, it's time to give NixOS another shot as a server.
